nixos/default.nix

```diff
@@ -21,65 +33,107 @@ in {
+        (attrsets.recursiveUpdate (baseUnit username) {
           wantedBy = [ "multi-user.target" ];
           wants = [ "nix-daemon.socket" ];
           after = [ "nix-daemon.socket" ];
           before = [ "systemd-user-sessions.service" ];
 
-          environment = serviceEnvironment;
+          unitConfig.RequiresMountsFor = homeDirectory;
+
+          serviceConfig.User = username;
+          serviceConfig.ExecStart = let
+            systemctl =
+              "XDG_RUNTIME_DIR=\${XDG_RUNTIME_DIR:-/run/user/$UID} systemctl";
+
+            sed = "${pkgs.gnused}/bin/sed";
+
+            exportedSystemdVariables = concatStringsSep "|" [
+              "DBUS_SESSION_BUS_ADDRESS"
+              "DISPLAY"
+              "WAYLAND_DISPLAY"
+              "XAUTHORITY"
+              "XDG_RUNTIME_DIR"
+            ];
+
+            setupEnv = pkgs.writeScript "hm-setup-env" ''
+              #! ${pkgs.runtimeShell} -el
+
+              # The activation script is run by a login shell to make sure
+              # that the user is given a sane environment.
+              # If the user is logged in, import variables from their current
+              # session environment.
+              eval "$(
+                ${systemctl} --user show-environment 2> /dev/null \
+                | ${sed} -En '/^(${exportedSystemdVariables})=/s/^/export /p'
+              )"
+
+              exec "$1/activate"
+            '';
+          in "${setupEnv} ${usercfg.home.activationPackage}";
+        })) cfg.users;
+    })
+    (mkIf (cfg.users != { } && cfg.useUserService) {
+      systemd.user.services.home-manager = (baseUnit "%u") // {
+        wantedBy = [ "default.target" ];
+
+        # user units cannot depend on system units
+        # TODO: Insert in the script logic for waiting on the nix socket via dbus
+        # like https://github.com/mogorman/systemd-lock-handler
+        # wants = [ "nix-daemon.socket" ];
+        # after = [ "nix-daemon.socket" ];
+
+        unitConfig.RequiresMountsFor = "%h";
+        # no ExecStart= is defined for any user that has not defined
+        # config.home-manager.users.${username}
+        # this will be overridden by the below drop-in
+      };
 
-          unitConfig = { RequiresMountsFor = usercfg.home.homeDirectory; };
+      users.users = mapAttrs (_: usercfg: {
+        packages =
+          # unit files are taken from $XDG_DATA_DIRS too
+          # but are loaded after units from /etc
+          # we write a drop in so that it will take precedence
+          # over the above unit declaration
+          [
+            (pkgs.writeTextDir "${hmDropIn}/10-user-activation.conf" ''
+              [Service]
+              ExecStart=${usercfg.home.activationPackage}/activate
+            '')
+          ];
+      }) cfg.users;
+      environment.pathsToLink = [ hmDropIn ];
 
-          stopIfChanged = false;
+      systemd.services = mapAttrs' (_: usercfg:
```

Comment:
Could this special casing for ssh be moved under a `mkIf (!config.programs.ssh.internallyManaged)`?

For my specific use-case (home directory mounted as `pam_mount`) trying to manage the ssh config seems strange when all changes are undone immediately upon login.

IMO doing stuff required for login (e.g. authorized keys) in the system config is the way to go.